src/coding_theory/hamming.lean

```diff
@@ -0,0 +1,245 @@
+hamm_dist_eq_zero_iff_forall_eq_zero
+
+lemma hamm_wt_ne_zero_iff_exists_ne_zero {x : Π i, β i} : hamm_wt x ≠ 0 ↔ ∃ i, x i ≠ 0 :=
+hamm_dist_ne_zero_iff_exists_ne_zero
+
+end has_zero
+
+lemma hamm_dist_eq_hamm_wt_sub [Π i, add_group (β i)] (x y : Π i, β i) :
+  hamm_dist x y = hamm_wt (x - y) :=
+by simp_rw [hamm_wt_eq, hamm_dist, pi.sub_apply, sub_ne_zero]
+
+end hamm_dist_wt
+
+/-! ### The `hamm` type synonym -/
+
+/--
+Type synonym for a Pi type which we equip with the Hamming metric, adding relevant
+instances as needed.
+-/
+def hamm {ι : Type*} (β : ι → Type*) : Type* := Π i, β i
+
+namespace hamm
+
+section
+
+variables {α ι : Type*} {β : ι → Type*}
+
+instance [Π i, inhabited (β i)] : inhabited (hamm β) := ⟨λ i, default⟩
+instance [decidable_eq ι] [fintype ι] [Π i, fintype (β i)] : fintype (hamm β) := pi.fintype
+instance [inhabited ι] [inst : ∀ i, nonempty (β i)] [nontrivial (β default)] : nontrivial (hamm β)
+:= pi.nontrivial
+instance [fintype ι] [Π i, decidable_eq (β i)] : decidable_eq (hamm β) :=
+fintype.decidable_pi_fintype
+instance [Π i, has_zero (β i)] : has_zero (hamm β) := pi.has_zero
+instance [Π i, has_add (β i)] : has_add (hamm β) := pi.has_add
+instance [Π i, has_neg (β i)] : has_neg (hamm β) := pi.has_neg
+instance [Π i, has_sub (β i)] : has_sub (hamm β) := pi.has_sub
+instance [Π i, has_scalar α (β i)] : has_scalar α (hamm β) := pi.has_scalar
+instance [has_zero α] [Π i, has_zero (β i)] [Π i, smul_with_zero α (β i)] :
+  smul_with_zero α (hamm β) := pi.smul_with_zero _
+instance [Π i, add_monoid (β i)] : add_monoid (hamm β) := pi.add_monoid
+instance [Π i, add_comm_monoid (β i)] : add_comm_monoid (hamm β) := pi.add_comm_monoid
+instance [Π i, add_comm_group (β i)] : add_comm_group (hamm β) := pi.add_comm_group
+instance (α) [semiring α] (β: ι → Type*) [Π i, add_comm_monoid (β i)]
+  [Π i, module α (β i)] : module α (hamm β) := pi.module _ _ _
+
+/-- `to_hamm` is the identity function to the `hamm` of a type.  -/
+@[pattern] def to_hamm : (Π i, β i) ≃ hamm β := equiv.refl _
+
+/-- `of_hamm` is the identity function from the `hamm` of a type.  -/
+@[pattern] def of_hamm : hamm β ≃ Π i, β i := equiv.refl _
+
+@[simp] lemma to_hamm_symm_eq                 : (@to_hamm _ β).symm = of_hamm := rfl
+@[simp] lemma of_hamm_symm_eq                 : (@of_hamm _ β).symm = to_hamm := rfl
+@[simp] lemma to_hamm_of_hamm (x : hamm β)    : to_hamm (of_hamm x) = x := rfl
+@[simp] lemma of_hamm_to_hamm (x : Π i, β i)  : of_hamm (to_hamm x) = x := rfl
+@[simp] lemma to_hamm_inj {x y : Π i, β i}    : to_hamm x = to_hamm y ↔ x = y := iff.rfl
+@[simp] lemma of_hamm_inj {x y : hamm β}      : of_hamm x = of_hamm y ↔ x = y := iff.rfl
+
+end
+
+section
+
+variables {α ι : Type*} {β : ι → Type*} [fintype ι] [Π i, decidable_eq (β i)]
+
+instance : has_dist (hamm β) := ⟨λ x y, hamm_dist (of_hamm x) (of_hamm y)⟩
+
+@[simp, push_cast] lemma dist_eq_hamm_dist (x y : hamm β) : dist x y =
+  hamm_dist (of_hamm x) (of_hamm y) := rfl
```

Comment:
Breaking things at the `=` should be a last resort, as it makes the lemma harder to read
```suggestion
@[simp, push_cast] lemma dist_eq_hamm_dist (x y : hamm β) :
  dist x y = hamm_dist (of_hamm x) (of_hamm y) := rfl
```